packages/xsnap/makefiles/mac/xsnap.mk

```diff
@@ -0,0 +1,142 @@
+% : %.c
+%.o : %.c
+
+GOAL ?= debug
+NAME = xsnap
+ifneq ($(VERBOSE),1)
+MAKEFLAGS += --silent
+endif
+
+HERE = $(dir $(realpath $(firstword $(MAKEFILE_LIST))))
+MODDABLE = $(HERE)/../../moddable
+XS_DIR = $(MODDABLE)/xs
+BUILD_DIR = $(HERE)/../../build
+
+BIN_DIR = $(BUILD_DIR)/bin/mac/$(GOAL)
+INC_DIR = $(XS_DIR)/includes
+PLT_DIR = $(XS_DIR)/platforms
+SRC_DIR = $(XS_DIR)/sources
+TLS_DIR = ../../src
+TMP_DIR = $(BUILD_DIR)/tmp/mac/$(GOAL)/$(NAME)
+
+# MACOS_ARCH ?= -arch i386
+MACOS_ARCH ?=
+MACOS_VERSION_MIN ?= -mmacosx-version-min=10.7
+
+C_OPTIONS = \
+	-fno-common \
+	$(MACOS_ARCH) \
+	$(MACOS_VERSION_MIN) \
+	-DINCLUDE_XSPLATFORM \
+	-DXSPLATFORM=\"xsnap.h\" \
+	-DmxDebug=1 \
+	-DmxMetering=1 \
+	-DmxParse=1 \
+	-DmxRun=1 \
+	-DmxSloppy=1 \
+	-DmxSnapshot=1 \
+	-DmxRegExpUnicodePropertyEscapes=1 \
+	-I$(INC_DIR) \
+	-I$(PLT_DIR) \
+	-I$(SRC_DIR) \
+	-I$(TLS_DIR) \
+	-I$(TMP_DIR)
+ifneq ("x$(SDKROOT)", "x")
+	C_OPTIONS += -isysroot $(SDKROOT)
+endif
+ifeq ($(GOAL),debug)
+	C_OPTIONS += -g -O0 -Wall -Wextra -Wno-missing-field-initializers -Wno-unused-parameter
+else
+	C_OPTIONS += -O3
+endif
+
+LIBRARIES = -framework CoreServices
+
+LINK_OPTIONS = $(MACOS_VERSION_MIN) $(MACOS_ARCH)
+ifneq ("x$(SDKROOT)", "x")
+	LINK_OPTIONS += -isysroot $(SDKROOT)
+endif
+
+# C_OPTIONS += -fsanitize=address -fno-omit-frame-pointer
+# LINK_OPTIONS += -fsanitize=address -fno-omit-frame-pointer
+
+OBJECTS = \
+	$(TMP_DIR)/xsAll.o \
+	$(TMP_DIR)/xsAPI.o \
+	$(TMP_DIR)/xsArguments.o \
+	$(TMP_DIR)/xsArray.o \
+	$(TMP_DIR)/xsAtomics.o \
+	$(TMP_DIR)/xsBigInt.o \
+	$(TMP_DIR)/xsBoolean.o \
+	$(TMP_DIR)/xsCode.o \
+	$(TMP_DIR)/xsCommon.o \
+	$(TMP_DIR)/xsDataView.o \
+	$(TMP_DIR)/xsDate.o \
+	$(TMP_DIR)/xsDebug.o \
+	$(TMP_DIR)/xsDefaults.o \
+	$(TMP_DIR)/xsError.o \
+	$(TMP_DIR)/xsFunction.o \
+	$(TMP_DIR)/xsGenerator.o \
+	$(TMP_DIR)/xsGlobal.o \
+	$(TMP_DIR)/xsJSON.o \
+	$(TMP_DIR)/xsLexical.o \
+	$(TMP_DIR)/xsMapSet.o \
+	$(TMP_DIR)/xsMarshall.o \
+	$(TMP_DIR)/xsMath.o \
+	$(TMP_DIR)/xsMemory.o \
+	$(TMP_DIR)/xsModule.o \
+	$(TMP_DIR)/xsNumber.o \
+	$(TMP_DIR)/xsObject.o \
+	$(TMP_DIR)/xsPlatforms.o \
+	$(TMP_DIR)/xsProfile.o \
+	$(TMP_DIR)/xsPromise.o \
+	$(TMP_DIR)/xsProperty.o \
+	$(TMP_DIR)/xsProxy.o \
+	$(TMP_DIR)/xsRegExp.o \
+	$(TMP_DIR)/xsRun.o \
+	$(TMP_DIR)/xsScope.o \
+	$(TMP_DIR)/xsScript.o \
+	$(TMP_DIR)/xsSnapshot.o \
+	$(TMP_DIR)/xsSourceMap.o \
+	$(TMP_DIR)/xsString.o \
+	$(TMP_DIR)/xsSymbol.o \
+	$(TMP_DIR)/xsSyntaxical.o \
+	$(TMP_DIR)/xsTree.o \
+	$(TMP_DIR)/xsType.o \
+	$(TMP_DIR)/xsdtoa.o \
+	$(TMP_DIR)/xsre.o \
+	$(TMP_DIR)/$(NAME).o
+
+VPATH += $(SRC_DIR) $(TLS_DIR)
+
+build: $(MODDABLE)/README.md $(TMP_DIR) $(BIN_DIR) $(BIN_DIR)/$(NAME)
+
+$(MODDABLE)/README.md:
+	git submodule init
+	git submodule update
```

Comment:
This ensures the submodule is initialized without involving the user.

Reply:
@warner did you talk me out of putting that stanza in the critical path? Or was I just hesitant to do it?

Reply:
A bit of both, at the time I was concerned about the additional dependencies (`glib` at least, some unusual flavor of `make` on some platforms?) that would be needed to build the XS tools (`xsc`, `xsbug`), so I didn't think mainstream users should be impacted by the XS build process.

@kriskowal showed me that this new `xsnap` build process uses the XS `.c` files directly, without depending upon the XS tools to be compiled first, which removes those dependencies. So I think I'm comfortable with building them unconditionally. Although we still need to pay attention to our second-tier targets (ohai windows) and how much we need to update our getting-started docs to get their tools in place.

But a `git submodule` on the primary build path doesn't sound too bad to me.